athena-jdbc/src/main/java/com/amazonaws/connectors/athena/jdbc/saphana/SAPHANAMetadataHandler.java

```diff
@@ -0,0 +1,196 @@
+                        blockWriter.writeRows((Block block, int rowNum) -> {
+                            block.setValue(BLOCK_PARTITION_COLUMN_NAME, rowNum, partitionName);
+                            LOGGER.info("Adding partition {}", partitionName);
+                            //we wrote 1 row so we return 1
+                            return 1;
+                        });
+                    }
+                    while (resultSet.next() && queryStatusChecker.isQueryRunning());
+                }
+            }
+        }
+        catch (SQLException sqlException) {
+            throw new RuntimeException(sqlException.getErrorCode() + ": " + sqlException.getMessage(), sqlException);
+        }
+    }
+
+    @Override
+    public GetSplitsResponse doGetSplits(
+            final BlockAllocator blockAllocator, final GetSplitsRequest getSplitsRequest)
+    {
+        LOGGER.info("{}: Catalog {}, table {}", getSplitsRequest.getQueryId(), getSplitsRequest.getTableName().getSchemaName(), getSplitsRequest.getTableName().getTableName());
+        int partitionContd = decodeContinuationToken(getSplitsRequest);
+        Set<Split> splits = new HashSet<>();
+        Block partitions = getSplitsRequest.getPartitions();
+
+        // TODO consider splitting further depending on #rows or data size. Could use Hash key for splitting if no partitions.
+        for (int curPartition = partitionContd; curPartition < partitions.getRowCount(); curPartition++) {
+            FieldReader locationReader = partitions.getFieldReader(BLOCK_PARTITION_COLUMN_NAME);
+            locationReader.setPosition(curPartition);
+
+            SpillLocation spillLocation = makeSpillLocation(getSplitsRequest);
+
+            LOGGER.info("{}: Input partition is {}", getSplitsRequest.getQueryId(), locationReader.readText());
+
+            Split.Builder splitBuilder = Split.newBuilder(spillLocation, makeEncryptionKey())
+                    .add(BLOCK_PARTITION_COLUMN_NAME, String.valueOf(locationReader.readText()));
+
+            splits.add(splitBuilder.build());
+
+            if (splits.size() >= MAX_SPLITS_PER_REQUEST) {
+                //We exceeded the number of split we want to return in a single request, return and provide a continuation token.
+                return new GetSplitsResponse(getSplitsRequest.getCatalogName(), splits, encodeContinuationToken(curPartition));
+            }
+        }
+
+        return new GetSplitsResponse(getSplitsRequest.getCatalogName(), splits, null);
+    }
+
+    private int decodeContinuationToken(GetSplitsRequest request)
+    {
+        if (request.hasContinuationToken()) {
+            return Integer.valueOf(request.getContinuationToken());
+        }
+
+        //No continuation token present
+        return 0;
+    }
+
+    private String encodeContinuationToken(int partition)
+    {
+        return String.valueOf(partition);
```

Comment:
don't you also need to know what split you left off at? since it is possible that you didn't finish the partition before exceeding the page size MAX_SPLITS_PER_REQUEST? the total response can not exceed ~6MB

Reply:
any progress on this?

Reply:
not sure, if i am getting you here. SAP HANA JDBC driver is mostly based on mySQL code with customisation for SAP HANA. This is being managed the same way in mySQL as well. Beside if i am understanding the flow correctly. We handle the split in GetSplitsResponse.